app/Http/Controllers/Rest/UsersController.php

```diff
@@ -0,0 +1,152 @@
+        Mail::to($user)->queue(
+            new UserEmailActivate(
+                Token::generate('email-activate', $this->email_activation_token_time, $user->id),
+                $request->input('uri')
+            )
+        );
+        DB::commit();
+
+        return self::json($user->toArray(), 201)
+            ->header('Location', route('app.users.read', ['user_id' => $user->id]));
+    }
+
+    public function count(Request $request)
+    {
+        throw new ServerErrorException('Not implemented yet');
+    }
+
+    public function browse(Request $request)
+    {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
+        // TODO check privileges
+        $users = User::with('AdserverWallet')->get();
+
+        return self::json($users->toArray());
+    }
+
+    public function delete(Request $request, $user_id)
+    {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
+        // TODO check privileges
+        // TODO reset email
+        // TODO process
+        $user = User::findOrFail($user_id);
+        $user->delete();
+
+        return self::json([], 204);
+    }
+
+    public function edit(Request $request)
+    {
+        if (!Auth::check() && !$request->has('user.token')) {
+            return self::json([], 401, ['message' => 'Required authenticated access or token authentication']);
+        }
+
+        DB::beginTransaction();
+        if (Auth::check()) {
+            $user = Auth::user();
+            $token_authorization = false;
+        } else {
+            if (false === $token = Token::check($request->input('user.token'), null, 'password-recovery')) {
+                DB::rollBack();
+
+                return self::json([], 422, ['message' => 'Authentication token is invalid']);
+            }
+            $user = User::findOrFail($token['user_id']);
+            $token_authorization = true;
+        }
+
+        $this->validateRequestObject($request, 'user', User::$rules);
+        $user->fill($request->input('user'));
+
+        if (!$request->has('user.password_new')) {
+            $user->save();
+            DB::commit();
+
+            return self::json($user->toArray(), 200);
+        }
+
+        if ($token_authorization) {
+            $user->password = $request->input('user.password_new');
+            $user->save();
+            DB::commit();
+
+            return self::json($user->toArray(), 200);
+        }
+
+        if (!$request->has('user.password_old') || !$user->validPassword($request->input('user.password_old'))) {
+            DB::rollBack();
+
+            return self::json($user->toArray(), 422, ['password_old' => 'Old password is not valid']);
+        }
+
+        $user->password = $request->input('user.password_new');
+        $user->save();
+        DB::commit();
+
+        return self::json($user->toArray(), 200);
+    }
+
+    public function read(Request $request, $user_id)
+    {
+        // TODO check privileges
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)